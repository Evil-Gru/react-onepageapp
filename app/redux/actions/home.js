/*
 * Home Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  USER_LIST_LOADING,
  USER_LIST_LOADING_SUCCESS,
  USER_LIST_LOADING_ERROR,
} from '../actionTypes/home';

export function userListLoadingAction() {
  return {
    type: USER_LIST_LOADING,
  };
}

export function userListLoadingSuccessAction(users) {
  return {
    type: USER_LIST_LOADING_SUCCESS,
    payload: {
      users,
    },
  };
}

export function userListLoadingErrorAction(error) {
  return {
    type: USER_LIST_LOADING_ERROR,
    error,
  };
}
