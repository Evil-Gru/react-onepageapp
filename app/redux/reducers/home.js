/*
 * HomeReducer
 *
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 *
 */

import produce from 'immer';
import {
  USER_LIST_LOADING,
  USER_LIST_LOADING_SUCCESS,
  USER_LIST_LOADING_ERROR,
} from '../actionTypes/home';

// The initial state of the App
export const initialState = {
  users: [],
  loading: false,
  error: false,
};

/* eslint-disable default-case, no-param-reassign */
const homeReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case USER_LIST_LOADING:
        draft.loading = true;
        draft.error = false;
        break;
      case USER_LIST_LOADING_SUCCESS:
        draft.loading = false;
        draft.error = false;
        draft.users = [...action.payload.users];
        break;
      case USER_LIST_LOADING_ERROR:
        draft.loading = false;
        draft.error = true;
    }
  });

export default homeReducer;
