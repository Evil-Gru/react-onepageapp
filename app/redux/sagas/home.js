/**
 * Gets the repositories of the user from Github
 */

import { call, put, takeLatest } from 'redux-saga/effects';

import ApiEndpoint from 'utils/api';
import request from 'utils/request';

import { LOAD_USER } from '../actionTypes/app';
import { USER_LIST_LOADING } from '../actionTypes/home';

import { loadUserSuccessAction, loadUserErrorAction } from '../actions/app';

import {
  userListLoadingSuccessAction,
  userListLoadingErrorAction,
} from '../actions/home';

const api = new ApiEndpoint();

export function* getCurrentUser({ payload: { userId } }) {
  // Select username from store
  const requestURL = api.getUserByIdPath(userId);

  try {
    const user = yield call(request, requestURL);
    yield put(loadUserSuccessAction(user.data));
  } catch (err) {
    yield put(loadUserErrorAction(err));
  }
}

export function* getUsers() {
  /**
   * get only 15 users
   */

  const pageNum = 1;
  const perPage = 15;

  const requestURL = api.getUserListPath(pageNum, perPage);

  try {
    const users = yield call(request, requestURL);

    users.data.sort((a, b) => {
      const nameA = `${a.first_name.toUpperCase()} ${a.last_name.toUpperCase()}`; // ignore upper and lowercase
      const nameB = `${b.first_name.toUpperCase()} ${b.last_name.toUpperCase()}`; // ignore upper and lowercase
      if (nameA > nameB) {
        return -1;
      }
      if (nameA < nameB) {
        return 1;
      }
      // names must be equal
      return 0;
    });

    yield put(userListLoadingSuccessAction(users.data));
  } catch (err) {
    yield put(userListLoadingErrorAction(err));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export default function* rootSaga() {
  // Watches for USER_LIST_LOADING actions and calls getUsers when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(USER_LIST_LOADING, getUsers);
  yield takeLatest(LOAD_USER, getCurrentUser);
}
