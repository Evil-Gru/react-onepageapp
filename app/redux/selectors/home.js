/**
 * Homepage selectors
 */

import { createSelector } from 'reselect';
import { initialState } from '../reducers/home';

const selectHome = state => state.home || initialState;

const makeSelectLoading = () =>
  createSelector(
    selectHome,
    homeState => homeState.loading,
  );

const makeSelectError = () =>
  createSelector(
    selectHome,
    homeState => homeState.error,
  );

const makeSelectUserList = () =>
  createSelector(
    selectHome,
    homeState => homeState.users,
  );

export { selectHome, makeSelectLoading, makeSelectError, makeSelectUserList };
