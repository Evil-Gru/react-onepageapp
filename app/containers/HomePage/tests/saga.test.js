/**
 * Tests for HomePage sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { USER_LIST_LOADING } from 'redux/actionTypes/home';

import {
  userListLoadingSuccessAction,
  userListLoadingErrorAction,
} from 'redux/actions/home';

import rootSaga, { getUsers } from 'redux/sagas/home';

/* eslint-disable redux-saga/yield-effects */
describe('getRepos Saga', () => {
  let getUsersGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getUsersGenerator = getUsers();

    const selectDescriptor = getUsersGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getUsersGenerator.next([]).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the userListLoadingSuccessAction if it requests the data successfully', () => {
    const response = [];
    const putDescriptor = getUsersGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(userListLoadingSuccessAction(response)));
  });

  it('should call the repoLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getUsersGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(userListLoadingErrorAction(response)));
  });
});

describe('githubDataSaga Saga', () => {
  const sagaIterator = rootSaga();

  it('should start task to watch for USER_LIST_LOADING action', () => {
    const takeLatestDescriptor = sagaIterator.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(USER_LIST_LOADING, getUsers),
    );
  });
});
