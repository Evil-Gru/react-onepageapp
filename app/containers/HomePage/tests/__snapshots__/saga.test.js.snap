// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRepos Saga should call the repoLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getRepos Saga should call the repoLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://api.github.com/users/mxstbr/repos?type=all&sort=updated",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRepos Saga should dispatch the reposLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getRepos Saga should dispatch the reposLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://api.github.com/users/mxstbr/repos?type=all&sort=updated",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRepos Saga should dispatch the userListLoadingSuccessAction if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://reqres.in/api/users?page=1&per_page=15",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRepos Saga should dispatch the userListLoadingSuccessAction if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "error": [TypeError: Cannot read property 'sort' of undefined],
      "type": "boilerplate/Home/USER_LIST_LOADING_ERROR",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;
